official/recommendation/data_download.py

```diff
@@ -0,0 +1,297 @@
+
+  Args:
+    ratings: A pandas DataFrame of the rating dataset.
+    sort: A boolean to indicate whether to sort the data based on timestamp.
+
+  Returns:
+    ratings: The processed DataFrame.
+  """
+  ratings['timestamp'] = pd.to_datetime(ratings['timestamp'], unit='s')
+  if sort:
+    ratings.sort_values(by='timestamp', inplace=True)
+  _describe_ratings(ratings)
+  return ratings
+
+
+def load_ml_1m(file_name, sort=True):
+  """Load the ml-1m dataset.
+
+  Args:
+    file_name: A string of the file name to be loaded.
+    sort: A boolean to indicate whether to sort the data based on timestamp.
+
+  Returns:
+    A processed pandas DataFrame of the rating dataset.
+  """
+  names = ['user_id', 'item_id', 'rating', 'timestamp']
+  ratings = pd.read_csv(file_name, sep='::', names=names, engine='python')
+  return process_movielens(ratings, sort=sort)
+
+
+def load_ml_20m(file_name, sort=True):
+  """Load the ml-20m dataset.
+
+  Args:
+    file_name: A string of the file name to be loaded.
+    sort: A boolean to indicate whether to sort the data based on timestamp.
+
+  Returns:
+    A processed pandas DataFrame of the rating dataset.
+  """
+  ratings = pd.read_csv(file_name)
+  ratings['timestamp'] = pd.to_datetime(ratings['timestamp'], unit='s')
+  names = {'userId': 'user_id', 'movieId': 'item_id'}
+  ratings.rename(columns=names, inplace=True)
+  return process_movielens(ratings, sort=sort)
+
+
+def load_file_to_df(file_name, sort=True):
+  """Load rating dataset into DataFrame.
+
+  Two data loading functions are defined to handle dataset ml-1m and ml-20m,
+  as they are provided with different formats.
+
+  Args:
+    file_name: A string of the file name to be loaded.
+    sort: A boolean to indicate whether to sort the data based on timestamp.
+
+  Returns:
+    A pandas DataFrame of the rating dataset.
+  """
+  dataset_name = os.path.basename(file_name).split('.')[0]
+  # ml-1m with extension .dat
+  file_extension = '.dat'
+  func = load_ml_1m
+  if dataset_name == 'ml-20m':
+    file_extension = '.csv'
+    func = load_ml_20m
+  ratings_file = os.path.join(file_name, 'ratings' + file_extension)
+  return func(ratings_file, sort=sort)
+
+
+def generate_train_eval_data(df, original_users, original_items):
+  """Generate the dataset for model training and evaluation.
+
+  Given all user and item interaction information, for each user, first to sort
+  the interactions based on timestamp. Then the latest one is taken out as
+  Test ratings (leave-one-out evaluation) and the remaining data for training.
+  The Test negatives are randomly sampled from all non-interacted items, and the
+  number of Test negatives is 100 (defined as NUMBER_NEGATIVES).
+
+  Args:
+    df: The DataFrame of ratings data.
+    original_users: The original unique user ids in the dataset.
+    original_items: The original unique item ids in the dataset.
+
+  Returns:
+    all_ratings: A set of the [user_id, item_id] with interactions.
```

Comment:
why a set of tuple with 2 items, and not just a dict? or a list with 2 items on each line, which is consistent with the other 2 return params?

Reply:
Yes, I explicitly convert it to a list before return.